Entity error rate

aro-max opened the pull request:
Implements a new metric called the weighted average bag of entity error rate: WA_BEER as discussed in [#115 ](https://github.com/ebu/benchmarkstt/issues/115)

The metric is implemented as an extension of --wer:  
--beer entities.json -o json
with entities.json file defined as : 
{ "entitie_1": w_1, "entitie_2" : w_2... }

entitie_n being the entity n, defined as a list of words
the w_1 is the weight defining the importance of entity 1 for the computation of the weighted average beer, WA_BEER.

The output is a json file : 

{ entity_1: {'beer':  x1, 'occurrence_ref': n1}
entity_2: {'beer': x2, 'occurrence_ref': n2}
entity_3: {'beer': x3, 'occurrence_ref': n3}
w_av_beer: {'beer': wa_1, 'occurrence_ref': n1+n2+n3} }

with:

entity_i : text string defining entity_1
x_i : is the beer of entity i
n_i : is the number of occurrences of entity_i in the reference file
w_av_beer is the weighted average beer

A new test is implemented.

Commit:
Add of a new metric to compute the entity error rate. This error rate is computed with a bag of word approach. The position of the entity does not change the metric. There are two options, 'bag_of_words' and 'weighted_bag_of_words'. With the  weighted_bag_of_words the error rate is weighted by entity. For bag_of_words the bag of entities is defined by a .txt or a .json file. For the weighted_bag_of_word the entities are defined on .json file.
Exemple of usage : see the tutorial
benchmarkstt --reference qt_subs.xml --reference-type plaintext --hypothesis qt_kaldi_hypothesis.txt --config config.conf --beer  entities.json weighted_bag_of_words

aro-max committed:
Define the weighted bag of entities error rate
as
Av_WBEER ([entity_1, ... entity_N) =
    w_1*BEER (entity_1)*L1/L ... w_N*BEER (entity_1))*LN/L
with
L1 =  number of occurences of entity 1 in the reference document
L = L1 + ... + LN
and
w_1 + ... + w_N = 1

added setter and getter to the BEER class

aro-max committed:
Added the tests for a WA BEER metric defined in class BEER in
metrics/core.py
make test passed

aro-max committed:
Deleted the mode argument in the BEER class: by default the BEER's
compare method is computing the weighted average bag of entity error
rate.
To use average bag of entity error rate just set the weight to equal
values.

aro-max committed:
Test the weighted average bag of entity error rate defined in
metrics/core.py
make test passed

aro-max committed:
corrects one typo : occurence replaced by occurrence

aro-max committed:
Adds the bag of entity error rate metric : --beer  entities.json

Example:
--------

benchmarkstt --reference qt_subs.xml --reference-type plaintext
--hypothesis qt_kaldi_hypothesis.txt --config config.conf
--beer  entities.json

The file entities.json defines the dictionary of entities with a weight
of importance. The weights are normalised within the tool.
{ "entity_1":10, "entity_2" : 100, "entity_3" :10 }

aro-max committed:
Adds tests for BEER() class defined in metrics/core.py

aro-max committed:
Modifies the test on BEER() outputs to be compatible  with python 3.5.

aro-max committed:
Describes the input file format in the documentation of BEER()

aro-max committed:
Adapts the test for windows and python 3.5

danielthepope commented on src/benchmarkstt/metrics/core.py:
```diff
@@ -159,6 +160,160 @@ def compare(self, ref: Schema, hyp: Schema) -> OpcodeCounts:
+    ne_hyp = number of detection of the entity in the hypothesis file
+    ne_ref = number of detection of the entity in the reference file
+
+                      abs(ne_hyp - ne_ref)
+    BEER (entity) =  ---------------------
+                           ne_ref
```
These formulas aren't showing up in the documentation as you expect:
<img width="694" alt="image" src="https://user-images.githubusercontent.com/5173183/99433605-9ff2d900-2905-11eb-8ce6-f363cff5a3c6.png">

In order to make them appear as preformatted text, the previous line needs to have two colons `::` and the preformatted section has to be indented. See the pydoc for WER as an example (core.py line 78)

aro-max replied:
Thanks, doc reformated and corrected

danielthepope commented on src/benchmarkstt/metrics/core.py:
```diff
@@ -159,6 +160,160 @@ def compare(self, ref: Schema, hyp: Schema) -> OpcodeCounts:
+
+    { "entity_1":W1, "entity_2" : W2, "entity_3" :W3 .. }
+
+    W_n being the non-normalized weight, the noramlisation of the weights is performed by the tool as :
```
Typo: "noramlisation" -> "normalization"

aro-max replied:
Thanks

danielthepope commented on src/benchmarkstt/metrics/core.py:
```diff
@@ -159,6 +160,160 @@ def compare(self, ref: Schema, hyp: Schema) -> OpcodeCounts:
+        self._weight = []
+        self._entities = []
+
+        if entities_file.endswith('.json'):
```
- If the file doesn't end with '.json', but is a JSON file, what would we expect to happen? I think it should try to parse the file anyway, so this `if` might not be required.
- And if the file isn't correctly formatted (whether or not it's a .json file), what would we expect then?

aro-max replied:
This is an option, the file is parsed as a JSON file whatever the file extension. If no file is passed as parameter then there are 2 options:
- generate an error message: the programme fails
- generate an error message with a default value 
Input file error
beer
====
w_av_beer: {'beer': 0, 'occurrence_ref': 0}

aro-max replied:
I pushed the second option:
Input file error
beer
====
w_av_beer: {'beer': 0, 'occurrence_ref': 0}

danielthepope commented on src/benchmarkstt/metrics/core.py:
```diff
@@ -159,6 +160,160 @@ def compare(self, ref: Schema, hyp: Schema) -> OpcodeCounts:
+                data = json.load(f)
+            self._entities = list(data.keys())
+            weight = list(data.values())
+            weight = [0 if w < 0 else w for w in weight]
+
+            # normalized the sum of the weights to 1
+            # after assignment when the file is not red
+            sw = sum(weight)
+            if sw > 0:
+                self._weight = [w / sw for w in weight]
```
There is duplicated logic here and in `set_weight`. Can the code in this constructor call `self.set_weight(weight)` instead?

aro-max replied:
Agree, change done

danielthepope commented on tests/benchmarkstt/test_metrics_core.py:
```diff
@@ -39,3 +39,29 @@ def test_wer(a, b, exp):
+    ['theresa may is here', 'theresa may is there', ['theresa may', 'here'], [100, 10], (0.045, 2)],
+    ['aa bb cc dd', 'aa bb cc dd', ['aa', 'bb'], [100, 10], (0.0, 2)],
+    ['aa bb cc dd', 'aa bb cc dd', ['aa', 'bb'], [10, 100], (0.0, 2)],
+    ['aa bb cc dd', 'aa bb cc dd ee ff gg hh', ['aa', 'ee'], [10, 100], (0.0, 1)],
```
Perhaps I don't understand the algorithm, but I think this example should have a BEER of >0. One string contains `ee` but the other doesn't.

aro-max replied:
The entities 'aa' and 'bb' are both one time in the reference and in the hypothesis so the BEER is 0 and the number of occurrences is 2.

danielthepope reviewed (changes requested):
Thanks for your contribution, and sorry it's taken me so long to review this PR.

I haven't carefully reviewed how the algorithm works since you're going to help me out with it soon. Thanks for the offer!

I've added some comments (particularly in the documentation) that should be addressed. You should be able to build the documentation yourself, by going to the project root and entering `make docs`. You can then view the generated files by changing to the `cd docs/build/html` directory and using `python3 -m http.server`.

aro-max committed:
Minor changes in response to comment for the pull request #146
- added an Error message if the input file is incorrect
- use set_weight() in the opcode
- changed the comments of the class BEER to generate the doc

aro-max committed:
Enriched the test coverage for BEER class in test_beer

danielthepope commented on src/benchmarkstt/metrics/core.py:
```diff
@@ -159,6 +160,157 @@ def compare(self, ref: Schema, hyp: Schema) -> OpcodeCounts:
+        return idx_found
+
+    # generate a list containing the detected entities in list_parsed
+    @staticmethod
+    def __generate_list_entity(self, list_parsed):
```
My editor is getting confused by this (it thinks `__find_pattern` is never accessed). If a function contains `self` as a parameter, it can't be a `@staticmethod`. You should be able to just remove the `@staticmethod` line.

aro-max replied:
It is accessed here:
index_entity = self.__find_pattern(list_parsed, entity)
Ok removed static method

danielthepope commented on src/benchmarkstt/metrics/core.py:
```diff
@@ -159,6 +160,157 @@ def compare(self, ref: Schema, hyp: Schema) -> OpcodeCounts:
+        return list_entity
+
+    # computes the BEER
+    @staticmethod
```
This `@staticmethod` should be removed because of the `self` in the function definition

danielthepope commented on src/benchmarkstt/metrics/core.py:
```diff
@@ -159,6 +160,157 @@ def compare(self, ref: Schema, hyp: Schema) -> OpcodeCounts:
+        # the cursor sweep complex_entity to find consecutive entities entity1 ... entity2
+        cursor = 0
+        idx_found = []
+        for idx, elt in enumerate(search_list):
```
This `idx` isn't being used. To avoid confusion with the definition of `idx` in line 252, line 248 can be simplified
```suggestion
        for elt in search_list:
```

aro-max replied:
No, it is not equivalent. I renamed the var idx to idex_sl for sake of readability but it can not be removed see line 252

danielthepope replied:
You're right. Thanks for clarifying.

danielthepope commented on src/benchmarkstt/metrics/core.py:
```diff
@@ -159,6 +160,157 @@ def compare(self, ref: Schema, hyp: Schema) -> OpcodeCounts:
+        self._weight = []
+        self._entities = []
+
+        try:
+            with open(entities_file) as f:
+                data = json.load(f)
+            self._entities = list(data.keys())
+            weight = list(data.values())
+            self.set_weight(weight)
+        except IOError:
+            print('Input file Error')
```
Printing the error message isn't the right thing to do here. One of benchmarkstt's output formats is `json`, which will print the results in a machine-parseable way.

`benchmarkstt --reference-type argument --hypothesis-type argument --reference 'the quick brown fox jumps over the lazy dog' --hypothesis 'the quick brown fax jumps over the lazy dog' --lowercase --wer --beer doesNotExist.json -o json`

This will output (on STDOUT)
```json
[
	{"title": "wer", "result": 0.1111111111111111}Input file Error
,
	{"title": "beer", "result": {"w_av_beer": {"beer": 0, "occurrence_ref": 0}}}
]
```
...which is invalid JSON, so will cause further problems if this tool were to be used programmatically.

It would in fact be better to remove the `try/except` completely, therefore we assume that the file name was written correctly and that it exists. If there are any errors, the stack trace will be printed to the screen on STDERR, which will be somewhat ugly, but at least STDOUT will still be valid JSON.

```suggestion
        with open(entities_file) as f:
            data = json.load(f)
        self._entities = list(data.keys())
        weight = list(data.values())
        self.set_weight(weight)
```

An alternative approach I can think of will be more complex. Instead of loading the JSON file in `__init__()`, just record the name of the file, so `self._entities_file = entities_file`. Then later on, you can load that file inside the `compare()` function, within a `try/except`. If an `IOError` or `json.decoder.JSONDecodeError` is thrown, you can choose to `return` the error message which will be formatted correctly no matter what output format is chosen.

aro-max replied:
Originally it generated an empty dict in case of invalid file, for sake of readability I put the reading of the file in the __init__(). If I move it to compare() it generates issues for testing. I would suggest generating an error that break the pipe or an empty dict as done in the first version.

danielthepope commented on src/benchmarkstt/metrics/core.py:
```diff
@@ -159,6 +160,157 @@ def compare(self, ref: Schema, hyp: Schema) -> OpcodeCounts:
+        hyp_list = [i['item'] for i in hyp]
+
+        # extract the entities
+        list_hypothesis_entity = self.__generate_list_entity(self, hyp_list)
+        list_reference_entity = self.__generate_list_entity(self, ref_list)
+        # compute the score
+        wer_entity = self.compute_beer(self, list_hypothesis_entity, list_reference_entity)
```
Since we can change `__generate_list_entity` and `compute_beer` so that they aren't `@staticmethod`, you can remove the `self` from the function call's parameters.

```suggestion
        list_hypothesis_entity = self.__generate_list_entity(hyp_list)
        list_reference_entity = self.__generate_list_entity(ref_list)
        # compute the score
        wer_entity = self.compute_beer(list_hypothesis_entity, list_reference_entity)
```

aro-max replied:
ok done

danielthepope commented on src/benchmarkstt/metrics/core.py:
```diff
@@ -159,6 +160,157 @@ def compare(self, ref: Schema, hyp: Schema) -> OpcodeCounts:
+            if elt == entity[cursor]:
+                cursor += 1
+                if cursor == le:
+                    idx_found.append([idx for idx in range(idx - le + 1, idx - le + 1 + le)])
```
This could be simplified a little bit
```suggestion
                    idx_found.append(list(range(idx - le + 1, idx - le + 1 + le)))
```

aro-max replied:
It can not be simplified this way. I renamed the idx of the upper level as idex_sl.

danielthepope replied:
OK. Now that the upper level has been named `idx_sl` I think that line can be changed to `idx_found.append(list(range(idx_sl - le + 1, idx_sl - le + 1 + le)))`. This just removes an extra variable definition (`idx`) that will make the code easier to understand.
I've run it with the new suggested line and all the tests still pass.

aro-max replied:
done

danielthepope commented on tests/benchmarkstt/test_metrics_core.py:
```diff
@@ -39,3 +39,32 @@ def test_wer(a, b, exp):
+
+    assert set(out.keys()) == set(entities_list)
+    assert out['w_av_beer']['beer'] == exp[0]
+    assert out['w_av_beer']['occurrence_ref'] == exp[1]
```
OK, now that I know what to expect from the BEER class, I think it would also be a good idea to add a unit test to demonstrate the other fields that are part of the output. At the moment we don't have any tests that check the values returned for a specific entity e.g. "madam". I would expect at least one test that should contain a line like `assert out['madam']['beer'] == 0.5`, but I'd recommend that this be a separate unit test, rather than trying to add complexity to test_beer.

aro-max replied:
ok I would propose to check the list of entities but the WA_BEER value is already a check of the BEER calculation.

aro-max replied:
I added test_beer() to test the beer per entity.

danielthepope reviewed (changes requested):
Alright! I understand how the algorithm works and how it calculates those numbers!

I've added a few suggestions. With regard to the "Input file error", I've suggested a different way to add a helpful error message, but this may be complicated to implement. I'd also be fine with throwing the error and stack trace to the console instead.

Thanks for addressing the previous comments :)

aro-max committed:
Remove @staticmethod, change naming of variable, change the error
message if the input file is not consistant

aro-max committed:
Change the error management, now it will generate a dictionary
containing the error message.
Use a setter to set the entities for testing purpose

aro-max committed:
test_beer() : test the beer per entity
test_wa_beer : test the weighted average beer

aro-max committed:
in __find_pattern() change one line for sake of readability

aro-max commented:
> Alright! I understand how the algorithm works and how it calculates those numbers!
> 
> I've added a few suggestions. With regard to the "Input file error", I've suggested a different way to add a helpful error message, but this may be complicated to implement. I'd also be fine with throwing the error and stack trace to the console instead.
> 
> Thanks for addressing the previous comments :)

I did the change in the code, now the error in jthe son output file

aro-max commented:
close by mistake

danielthepope reviewed (approved):
This is excellent, thanks very much for persevering with all my comments!

danielthepope committed:
Merge branch 'master' into entity_error_rate
